### PR TITLE
feat(span): implement `CloneIn` for the AST-related items.

### PR DESCRIPTION
**Pull request by rzvxa**
Follow-on after #4276, related to #4284.

### CODE REVIEW

**Comment by rzvxa** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4729?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4729** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4729?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#4728** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4728?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4726** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4726?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rzvxa and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/oxc_span/src/atom.rs [hunk 1]
```diff
@@ -59,6 +59,14 @@ impl<'a> Atom<'a> {
     }
 }
 
+impl<'old_alloc, 'new_alloc> CloneIn<'new_alloc> for Atom<'old_alloc> {
+    type Cloned = Atom<'new_alloc>;
+
+    fn clone_in(&self, alloc: &'new_alloc Allocator) -> Self::Cloned {
+        Self::Cloned::from_in(self.as_str(), alloc)
```
Comment (overlookmotel) [inline crates/oxc_span/src/atom.rs:66]: I think this would be slightly clearer:

```suggestion
        Atom::from_in(self.as_str(), alloc)
```

**Comment by overlookmotel** [comment]
The benchmark result is spurious! CodSpeed being weird again.

**Comment by rzvxa** [comment]
> The benchmark result is spurious! CodSpeed being weird again.

yeah, it happened a few times for me working on this stack. I reordered them a bunch might be related.

**Review by overlookmotel — APPROVED** [review]

**Pull request merged**